src/tests/string_consumer.cpp

```diff
@@ -0,0 +1,419 @@
+	CHECK(reader.Read(2) == "");
+}
+
+TEST_CASE("StringConsumer - ascii")
+{
+	StringConsumer reader("abcdefgh  \r\n\tAB  \r\n\t"sv);
+	CHECK(reader.FindCharIn("dc") == 2);
+	CHECK(reader.FindCharIn("xy") == StringConsumer::npos);
+	CHECK(reader.FindCharNotIn("ba") == 2);
+	CHECK(reader.PeekUntilCharNotIn("ba") == "ab");
+	CHECK(reader.PeekUntilCharNotIn("dc") == "");
+	CHECK(reader.PeekUntilCharIn("ba") == "");
+	CHECK(reader.PeekUntilCharIn("dc") == "ab");
+	CHECK(reader.ReadUntilCharNotIn("dc") == "");
+	CHECK(reader.ReadUntilCharNotIn("ba") == "ab");
+	CHECK(reader.ReadUntilCharIn("dc") == "");
+	CHECK(reader.ReadUntilCharIn("fe") == "cd");
+	CHECK(reader.PeekIf("ef"));
+	reader.SkipUntilCharNotIn("ji");
+	CHECK(reader.PeekIf("ef"));
+	reader.SkipUntilCharNotIn("fe");
+	CHECK(reader.PeekIf("gh"));
+	reader.SkipUntilCharIn("hg");
+	CHECK(reader.PeekIf("gh"));
+	reader.SkipUntilCharIn(StringConsumer::WHITESPACE_OR_NEWLINE);
+	CHECK(reader.PeekCharIfIn(StringConsumer::WHITESPACE_OR_NEWLINE) == ' ');
+	CHECK(reader.ReadCharIfIn(StringConsumer::WHITESPACE_OR_NEWLINE) == ' ');
+	reader.SkipCharIfIn(StringConsumer::WHITESPACE_OR_NEWLINE);
+	CHECK(reader.PeekUntilCharNotIn(StringConsumer::WHITESPACE_NO_NEWLINE) == "\r");
+	CHECK(reader.ReadUntilCharNotIn(StringConsumer::WHITESPACE_NO_NEWLINE) == "\r");
+	reader.SkipUntilCharNotIn(StringConsumer::WHITESPACE_NO_NEWLINE);
+	CHECK(reader.PeekCharIfIn(StringConsumer::WHITESPACE_OR_NEWLINE) == '\n');
+	CHECK(reader.ReadCharIfIn(StringConsumer::WHITESPACE_OR_NEWLINE) == '\n');
+	CHECK(reader.PeekUntilCharNotIn(StringConsumer::WHITESPACE_NO_NEWLINE) == "\t");
+	CHECK(reader.ReadUntilCharNotIn(StringConsumer::WHITESPACE_NO_NEWLINE) == "\t");
+	reader.SkipUntilCharNotIn(StringConsumer::WHITESPACE_NO_NEWLINE);
+	CHECK(reader.PeekUntilCharIn(StringConsumer::WHITESPACE_OR_NEWLINE) == "AB");
+	CHECK(reader.ReadUntilCharIn(StringConsumer::WHITESPACE_OR_NEWLINE) == "AB");
+	CHECK(reader.PeekUntilCharNotIn(StringConsumer::WHITESPACE_OR_NEWLINE) == "  \r\n\t");
+	reader.SkipUntilCharNotIn(StringConsumer::WHITESPACE_OR_NEWLINE);
+	CHECK(!reader.AnyBytesLeft());
+}
+
+TEST_CASE("StringConsumer - parse int")
+{
+	StringConsumer reader("1 a -2 -2 ffffFFFF ffffFFFF -1aaaAAAA -1aaaAAAA +3 1234567890123 1234567890123 1234567890123 ffffFFFFffffFFFE ffffFFFFffffFFFE ffffFFFFffffFFFE ffffFFFFffffFFFE -0x1aaaAAAAaaaaAAAA -1234567890123 "sv);
+	CHECK(reader.PeekIntegerBase<uint32_t>(10) == std::pair<StringConsumer::size_type, uint32_t>(1, 1));
+	CHECK(reader.PeekIntegerBase<int32_t>(10) == std::pair<StringConsumer::size_type, int32_t>(1, 1));
+	CHECK(reader.TryReadIntegerBase<uint32_t>(10) == 1);
+	CHECK(reader.ReadUtf8() == ' ');
+	CHECK(reader.PeekIntegerBase<uint32_t>(16) == std::pair<StringConsumer::size_type, uint32_t>(1, 0xa));
+	CHECK(reader.PeekIntegerBase<int32_t>(16) == std::pair<StringConsumer::size_type, int32_t>(1, 0xa));
+	CHECK(reader.ReadIntegerBase<uint32_t>(16) == 0xa);
+	CHECK(reader.ReadUtf8() == ' ');
```

Comment:
```suggestion
	CHECK(reader.TryReadIntegerBase<uint32_t>(10) == std::nullopt);
	CHECK(reader.ReadUtf8() == ' ');
```
I'm missing the check for base 10 reading when the number is definitely base 16.